File: socialview-commons/src/main/java/com/hendraanggrian/appcompat/widget/Mention.java
package com.hendraanggrian.appcompat.widget;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/** Simple optional mention data class, use when there is no custom mention class. */
public class Mention implements Mentionable {
  private final CharSequence username;
  private final CharSequence displayname;
  private final Object avatar;

  public Mention(@NonNull CharSequence username) {
    this(username, null);
  }

  public Mention(@NonNull CharSequence username, @Nullable CharSequence displayname) {
    this(username, displayname, null);
  }

  public Mention(
      @NonNull CharSequence username, @Nullable CharSequence displayname, @Nullable Object avatar) {
    this.username = username;
    this.displayname = displayname;
    this.avatar = avatar;
  }

  @Override
  public boolean equals(@Nullable Object obj) {
    return obj instanceof Mention && ((Mention) obj).username == username;
  }

  @Override
  public int hashCode() {
    return username.hashCode();
  }

  @NonNull
  @Override
  public String toString() {
    return username.toString();
  }

  /** {@inheritDoc} */
  @NonNull
  @Override
  public CharSequence getUsername() {
    return username;
  }

  /** {@inheritDoc} */
  @Nullable
  @Override
  public CharSequence getDisplayname() {
    return displayname;
  }

  /** {@inheritDoc} */
  @Nullable
  @Override
  public Object getAvatar() {
    return avatar;
  }
}

File: socialview-commons/src/com/hendraanggrian/appcompat/widget/Mention.kt
package com.hendraanggrian.appcompat.widget

/** A data class of mention to be used with [com.hendraanggrian.widget.MentionAdapter]. */
data class Mention @JvmOverloads constructor(

    /** Unique id of this mention. */
    val username: String,

    /** Optional display name, located above username. */
    val displayname: String? = null,

    /** Optional avatar, may be Drawable, resources, or string url pointing to image. */
    val avatar: Any? = null
) {

    override fun equals(other: Any?): Boolean = other != null &&
        other is Mention &&
        other.username == username

    override fun hashCode(): Int = username.hashCode()

    override fun toString(): String = username
}